feat(releasing): add deprecation.d fragment system (#25442)
vectordotdev/vector: docs/DEPRECATION_POLICY.md

  release page automatically.
- Add a deprecation note to the component docs. Typically, this means adding `deprecation: "description of the deprecation"`
  to the `cue` data for the option or feature. If the `cue` schema does not support `deprecation` for whatever you
  are deprecating yet, add it to the schema and open an issue to have it rendered on the website.
- For a component that is being renamed, remove the documentation page for the old name and add a new one for the new
  name. Add an alias so the old name redirects. The title of the new name should be appended with the text
  `(formerly OldName)`.
- Add a `WARN`-level log message starting with the word `DEPRECATION` if Vector detects the deprecated configuration
  or feature being used (when possible).

When removing a deprecation in a subsequent release, the pull request should:

- Mark the change as breaking by including `!` in the title after the type/scope.
- Remove the deprecation from the component documentation.
- Add a changelog fragment with `type="breaking"` ([`changelog.d/README.md`](../changelog.d/README.md)). A short

>   one-line summary is sufficient — the enacted deprecation entry is the canonical record of what was removed.

Reviewer: Addressed in 115691da62.

When introducing a deprecation into Vector, the pull request introducing the deprecation should:

- **Add a deprecation fragment** to [`deprecation.d/`](../deprecation.d/) following the format in
  [`deprecation.d/README.md`](../deprecation.d/README.md). Set `deprecated_since` to the current release version.
  Use the fragment body for the full migration guide (rationale, before/after examples, links). Then run
  `cargo vdev deprecation generate` to regenerate `website/data/deprecations.json` and commit both files.
  Run `cargo vdev deprecation show` to preview, and `cargo vdev deprecation check` to validate.

> - Add a changelog fragment with `type="deprecation"` ([`changelog.d/README.md`](../changelog.d/README.md)). A short

Author: ```suggestion
- Add a deprecation changelog fragment in [`changelog.d`](../changelog.d/README.md). A short
```

A little shorter and besides `type="deprecation"` is not an accepted format

Reviewer: Applied the wording in b83712f891. Note: `deprecation` *is* a valid changelog fragment type — see `changelog.d/README.md` ("`deprecation`: A change that is introducing a deprecation.") and `scripts/check_changelog_fragments.sh` (`FRAGMENT_TYPES="breaking|security|deprecation|feature|enhancement|fix"`). It maps to the `chore` cue bucket in `vdev/src/commands/release/generate_cue.rs::parse_changelog_fragment`. Accepting the wording on its own merits.

>   one-line summary is sufficient — the deprecation fragment is the canonical migration guide and is rendered on the

Author: ```suggestion
  one-line summary is sufficient as the deprecation fragment is the canonical migration guide and is rendered on the
```

Reviewer: Applied in b83712f891 (same commit as the wording change above).

  release page automatically.

> - Add a deprecation note to the component docs. Typically, this means adding `deprecation: "description of the deprecation"`

Author: ```suggestion
- Add a deprecation note to the component docs if applicable. Typically, this means adding `deprecation: "description of the deprecation"`
  to the cue file or `#[configurable(deprecated = "use <alternative> instead")]` to the parameter.
```

Reviewer: Addressed in 8fefac80fc.

  to the `cue` data for the option or feature. If the `cue` schema does not support `deprecation` for whatever you
  are deprecating yet, add it to the schema and open an issue to have it rendered on the website.
- For a component that is being renamed, remove the documentation page for the old name and add a new one for the new
  name. Add an alias so the old name redirects. The title of the new name should be appended with the text
  `(formerly OldName)`.
- Add a `WARN`-level log message starting with the word `DEPRECATION` if Vector detects the deprecated configuration
  or feature being used (when possible).

When removing a deprecation in a subsequent release, the pull request should:

- Mark the change as breaking by including `!` in the title after the type/scope.
- Remove the deprecation from the component documentation.
- Add a changelog fragment with `type="breaking"` ([`changelog.d/README.md`](../changelog.d/README.md)). A short

>   one-line summary is sufficient — the enacted deprecation entry is the canonical record of what was removed.

Author: Should we even include **_a_** changelog? This will be its own section in the release. I think we can fully remove this entry from this doc

Reviewer: Per offline discussion: the two surfaces answer different questions and a reader using only one of them shouldn't be left with a gap.

* changelog.d/ → Vector Changelog - Breaking changes should also be included in the changelog
* Enacted Deprecations is a new view that we offer for transparency and to show that we followed our own policy.

Note: exception for bug fixes (that break behavior) will not show up in the second view and that's by design.

  the new name will be appended with the text `(formerly OldName)`.
- Add a log message to Vector that is logged at the `WARN` level starting with the word `DEPRECATION` if Vector detects
  the deprecated configuration or feature being used (when possible).
- Add a deprecation notice file to [`deprecation.d/`](../deprecation.d/) following the format described in
  [`deprecation.d/README.md`](../deprecation.d/README.md). Set `deprecated_since` to the current release version.

>   Then run `cargo vdev deprecation generate` to regenerate `website/data/deprecations.json` and

Reviewer: Addressed in 4733eb45bc.
